ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs

```diff
@@ -0,0 +1,194 @@
+                     tabulate, (===))
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck (testProperty)
+
+import           Control.Monad.Class.MonadAsync (MonadAsync)
+import           Control.Monad.Class.MonadFork (MonadFork)
+import           Control.Monad.Class.MonadSTM (MonadSTM, STM, atomically)
+import           Control.Monad.Class.MonadThrow (MonadMask, MonadThrow)
+import           Control.Monad.IOSim (runSimOrThrow)
+
+import           Ouroboros.Consensus.Ledger.Abstract (ledgerConfigView)
+import           Ouroboros.Consensus.Mempool (Mempool (..),
+                     MempoolSnapshot (..), openMempool)
+import           Ouroboros.Consensus.Mempool.TxSeq (TicketNo)
+import           Ouroboros.Consensus.Util.ThreadRegistry (withThreadRegistry)
+
+import           Ouroboros.Network.Chain (Chain (..))
+
+import           Ouroboros.Storage.ChainDB.API (ChainDB)
+import qualified Ouroboros.Storage.ChainDB.API as ChainDB
+import qualified Ouroboros.Storage.ChainDB.Mock as Mock
+
+import           Test.Util.TestBlock (GenTx (..), GenTxId (..), TestBlock,
+                     computeGenTxId, singleNodeTestConfig, testInitExtLedger)
+import           Test.Util.TestTx (TestTx (..))
+
+tests :: TestTree
+tests = testGroup "Mempool"
+  [ testProperty "getTxs == getTxsAfter zeroIdx"        prop_Mempool_getTxs_getTxsAfter
+  , testProperty "addedTxs == getTxs"                   prop_Mempool_addTxs_getTxs
+  , testProperty "Invalid transactions are never added" prop_Mempool_InvalidTxsNeverAdded
+  ]
+
+{-------------------------------------------------------------------------------
+  Mempool Implementation Properties
+-------------------------------------------------------------------------------}
+
+-- | Test that @getTxs == getTxsAfter zeroIdx@.
+prop_Mempool_getTxs_getTxsAfter :: Chain TestBlock
+                                -> [TestTx]
+                                -> Property
+prop_Mempool_getTxs_getTxsAfter bc txs =
+  testAddTxsWithMempoolAndSnapshot
+    bc
+    txs
+    (\Mempool{zeroIdx} MempoolSnapshot{getTxs, getTxsAfter} ->
+      getTxs === getTxsAfter zeroIdx)
+
+-- | Test that all valid transactions added to a 'Mempool' can be retrieved
+-- afterward.
+--
+-- n.b. In this test, we add a list of arbitrary transactions to an initially
+-- empty 'Mempool', retrieve all of the transactions from the 'Mempool', and
+-- assert that those transactions are /specifically/ equal to the /valid/
+-- transactions that we attempted to add (the invalid ones should have been
+-- dropped).
+prop_Mempool_addTxs_getTxs :: Chain TestBlock
+                           -> [TestTx]
+                           -> Property
+prop_Mempool_addTxs_getTxs bc txs =
+  testAddTxsWithMempoolAndSnapshot
+    bc
+    txs
+    (\_ MempoolSnapshot{getTxs} ->
+      filter (genTxIsValid . snd) (testTxsToGenTxPairs txs)
+        === dropThd getTxs)
+ where
```

Comment:
BTW, we're using 2 spaces for indentation, unless there's a `where`, in which case we use 4 spaces and indent `where` with 2 spaces. Let's stick to that for consistency.

Reply:
Oops.